docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
+A handy package to explore this graph is [torchviz](https://github.com/szagoruyko/pytorchviz).
+
+We use ``autograd.Function`` as a nice interface to create this graph.
+Indeed, during the ``.apply``, the forward is invoked and all the inputs and outputs are connected properly to the exiting graph associated to the inputs.
+The :code:`Node` that is created is a wrapper around the user-defined backward function (and the name is appended with ``Backward`` as you can see in the
+graph examples in torchviz.
+
+
+Not Gradients
+"""""""""""""
+
+The operations that do not compute gradients that are presented above are implemented as follows to obtain the behavior described above.
+
+Detach is returning a new Tensor object that shares the same Storage.
+The Tensor is the same except that it does not require gradient and its grad_fn is always ``None``.
+
+No grad simply prevents any :code:`Node` from being added to the graph.
+
+
+Inplace handling
+""""""""""""""""
+
+To get the required behavior for inplace operations, two things are needed.
+First, to be able to ensure correctness, we need to track the versions of the different Tensors to make sure that a :code:`Node` has access to the version it needs, not a modified one.
+Second, we need to build a graph that will perform the correct backward computation.
+
+From the python api, the current version of a Tensor can be queried with `._version`.
+The `Function` wrapper takes care of managing this version by using the information provided by the `.mark_dirty()` method available on the context.
+It also takes care of making sure that the Tensors given to the `.backward()` function have not been modified inplace.
+
+Building the graph in the absence of any views is simple as all the inplace operations can be "reinterpreted" as out of place.
+The inplace operations will only create a linear chain of :code:`Node`s as if `x.op_()` was writen `x = x.op()` (assuming corresponding out of place operation exists).
+
+Building the graph is more complex as we allow both views and inplace operations.
+Here, the view operations are responsible to mark any views that they make as being linked to the input.
+You can check if a Tensor is a view with `._is_view` and query the Tensor it is a view of with `._base`.
+Two case can happen then:
+
+- If this Tensor is not modified inplace, then its `.grad_fn` is the :code:`Node` associated with the view operation. And the base's `.grad_fn` is the :code:`Node` that corresponds to its forward function.
+- If this Tensor is modified inplace, its graph is "rebased". This has a few effects:
+
+  - The base `.grad_fn` becomes a `CopySlices`. This special :code:`Node` is a wrapper around the :code:`Node` corresponding to the inplace operations and apply it to the subset of the base on which the inplace operation was performed (through the view).
+  - All the views' `.grad_fn` become `AsStridedBackard`s that point to the newly created `CopySlices` from the base.
```

Comment:
nit: you probably want to use :code:`something` all over instead of just `something`.